src/com/perl5/lang/perl/idea/refactoring/PerlRefactor.java

```diff
@@ -0,0 +1,420 @@
+
+  private String codeSelection;
+  private String className = "";
+
+  private final Set<String> hashVariables = new HashSet<>();
+  private final Set<String> arrayVariables = new HashSet<>();
+  private final Set<String> scalarVariables = new HashSet<>();
+  private final Set<String> loopScalarVariables = new HashSet<>();
+  private final Set<String> localScalarVariables = new HashSet<>();
+  private final Set<String> localArrayVariables = new HashSet<>();
+  private final Set<String> localHashVariables = new HashSet<>();
+
+
+  private final ArrayList<String> parameters = new ArrayList<>();
+  private final ArrayList<String> innerVariables = new ArrayList<>();
+  private final ArrayList<String> outerVariables = new ArrayList<>();
+  private String methodParameters;
+  private String methodReturn;
+
+  /**
+   * @return The code selection passed in and modified from refactoring
+   */
+  public String getCodeSelection() {
+    return codeSelection;
+  }
+
+  /**
+   * @return An array of parameters to be passed in to a method
+   */
+  public ArrayList<String> getParameters() {
+    return parameters;
+  }
+
+  /**
+   * @return An array of parameters to be returned from a method
+   */
+  public ArrayList<String> getInnerVariables() {
+    return innerVariables;
+  }
+
+  /**
+   * @return returns an array of parameters with arrays and hashes as scalars.
+   */
+  public ArrayList<String> getOuterVariablesModified() {
+    String reg = "[@%](.*)";
+    Pattern regex = Pattern.compile(reg);
+    ArrayList<String> ovs = new ArrayList<>();
+
+    for (String s : this.outerVariables) {
+      Matcher matcher = regex.matcher(s);
+      if (matcher.find()) {
+        s = s.replaceAll(reg, "\\$" + matcher.group(1));
+      }
+      ovs.add(s);
+    }
+
+    return ovs;
+  }
+
+  public PerlRefactor(String codeSelection, String subName) {
+    this.codeSelection = codeSelection;
+    this.subName = subName;
+
+    parseVariables();
+    parseLocalVariables();
+    buildMethod();
+  }
+
+  /**
+   * @param methodName       Name of method to generate
+   * @param methodParameters parameters to pass into method
+   * @param methodBody       body of the generated method
+   * @param methodReturn     return string ie "return ($foo);"
+   * @return returns a built method
+   */
+  public String buildMethod(@NotNull String methodName,
+                            @Nullable AbstractVariableData[] methodParameters,
+                            @NotNull String methodBody,
+                            @Nullable String methodReturn) {
+    StringBuilder builder = new StringBuilder("sub ");
+    builder.append(methodName).append(" {\n");
+
+    if (methodParameters != null) {
+      builder.append(this.buildMethodParameters(methodParameters)).append("\n\n");
+    }
+
+    builder.append(methodBody).append("\n");
+
+    builder.append(methodReturn).append("\n");
+    builder.append("}\n");
+
+    return builder.toString();
+  }
+
+  // TODO: fix formatting ?
```

Comment:
it should be handled automatically

Reply:
I'm assuming you mean when switched to the psi tree approach?

Reply:
yep

Reply:
Feel free to contact me in Skype if you'll decide to dive in and will have questions. Think twice though. Takes time :)